스프링 시큐리티 및 JWT 사용을 위한 라이브러리 의존성 추가 및 기반 설정 작업

kmmin78 opened the pull request:
스프링 시큐리티 및 JWT 사용을 위한 라이브러리 의존성 추가 및 기반 설정 작업

주요 내용

1. 관련 라이브러리 의존성 추가 (spring security, spring boot test, jjwt, restAssured)
2. WebSecurity 설정 , app-core의 도메인 빈 설정.
  -> UserRepository를 빈 설정하기 위해 구현체 클래스 및 팩토리 메서드 생성

기존 PR(https://github.com/f-lab-edu/awesome-wiki/pull/53)은 닫고, 새로 작성한 PR입니다.

Closes #52

Commit:
주요 내용

로그인 기능을 도메인 domain UseCase 인터페이스, 테스트 코드로 작성

외부 프레임워크, 인프라 의존성이 많은 app-main 모듈에서 해당 인터페이스를 의존한다.

Closes #50

Commit:
스프링 시큐리티 및 JWT 사용을 위한 라이브러리 의존성 추가 및 기반 설정 작업

주요 내용

1. 관련 라이브러리 의존성 추가 (spring security, spring boot test, jjwt, restAssured)
2. WebSecurity 설정 , app-core의 도메인 빈 설정.
  -> UserRepository를 빈 설정하기 위해 구현체 클래스 및 팩토리 메서드 생성

Closes #52

FrancescoJo commented on app-main/src/main/kotlin/kr/flab/wiki/app/appconfig/security/WebSecurityConfig.kt:
```diff
@@ -0,0 +1,63 @@
+    private val authenticationProvider: AuthenticationProvider
+) : WebSecurityConfigurerAdapter() {
+    @Bean
+    override fun authenticationManagerBean(): AuthenticationManager {
```
요거, 왠만하면 Bean 선언만 모여있는 별도 파일로 빼는게 좀더 유지보수하기 좋을 것 같습니다.
`@Bean` 이 도처에 널리기 시작하면 나중에 그거 찾는것도 일이니까요.

`AuthenticationProvider` 타입을 주입받는게 일이긴 한데 `@Configuration` 으로 잘 주입받는거 보니 `@Bean` 선언으로 주입받아도 크게 문제는 없으리라 생각합니다.

kmmin78 replied:
이건 저번 pr에서도 질문을 드렸던 내용인데,
AuthenticationManager을 빈으로 생성하기 위해서는 WebSecurityConfigurerAdapter를 상속 받아야합니다.
WebSecurityConfigurerAdapter를 다른 Configuration 파일에서 또 상속을 받으면, 
다음과 같은 메세지가 뜨면서 스프링이 시작되지 않습니다.
@Order on WebSecurityConfigurers must be unique

짐작컨데, WebSecurityConfigurers는 하나의 Configuration에서만 상속을 받을 수 있다는 건데,
그렇게 되면 authenticationManagerBean은 해당 파일에서만 생성할 수 있게 됩니다.

다른 Bean 설정을 WebSecurityConfig.kt에 모으는 게 나을까요?
아니면, 그대로 유지하는 것 밖에 답이 없을까요?

FrancescoJo replied:
```
    override fun authenticationManagerBean(): AuthenticationManager {
        return super.authenticationManagerBean()
    }
```

이 구현을 호출하는 구문을 밖에서 만들어 주면 될 것 같은데요.

즉, 이렇게 구현해 보시면 어떨까요

```
@Configuration
@EnableWebSecurity
class WebSecurityConfig(
    private val authenticationProvider: AuthenticationProvider
) : WebSecurityConfigurerAdapter() {
    override fun authenticationManagerBean(): AuthenticationManager {
        return super.authenticationManagerBean()
    }

    // ...
}

@Configuration
class SomeBeanConfigs {
    @Bean
    override fun authenticationManager(wsConfig: WebSecurityConfig): AuthenticationManager {
        return wsConfig.authenticationManagerBean()
    }
}
```

FrancescoJo reviewed (changes requested):
`@Bean` 선언 외부화

kmmin78 committed:
Merge branch 'develop' into feature/kms/LoginWithSpringSecurityAndJWT/002

kmmin78 committed:
app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성 (#63)

* app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성

주요 내용

1. 시큐리티 인증 구현체 코드 작성 및 빈 설정
2. jwt (rs256) 생성 유틸클래스 작성
3. 시큐리티 로그인의 end-to-end 테스트 코드 작성

Closes #56

* app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성
및 추가 수정

주요 내용

1. 테스트 코드 method chain 간소화 (with rest-assured:kotlin-extensions)
2. email, password 지역변수 제거 (보안)
3. 테스트 코드 정상동작 확인 (@MockBean 활용)

Closes #56

* 코드리뷰 피드백 수정사항 반영

주요 내용

1. 코드리뷰 피드백 수정사항 반영

Closes #56

* 코드리뷰 피드백 수정사항 반영

주요 내용

1. 빈 설정 외부화
2. authentication.name, credentials null case 처리

Closes #56

Co-authored-by: kms <kmmin78@gmail.com>
